backend/app.py

```diff
@@ -216,39 +233,63 @@ def get_cset_members_items(codeset_ids: List[int] = None, con=CON) -> List[Legac
         {'codeset_ids': codeset_ids})
 
 
-def hierarchy(codeset_ids: List[int] = None, selected_concept_ids: List[int] = None) -> Dict:
-    """Get hierarchy of concepts in selected concept sets"""
-    if not selected_concept_ids:
-        selected_concept_ids = get_concept_set_member_ids(codeset_ids, column='concept_id')
-
-    # selected_roots: List[int] = top_level_cids(selected_concept_ids)
+def get_parent_children_map(root_cids: List[int], cids: List[int], con=CON) -> Dict[int, List[int]]:
+    """New hierarchy info from sql"""
+    # int(x): Prevents SQL injection by throwing error if not an integer
+    root_cids: List[int] = [int(x) for x in root_cids]
+    cids: List[int] = [int(x) for x in cids]
+    root_cids: str = ', '.join([str(x) for x in root_cids])
+    cids: str = ', '.join([str(x) for x in cids])
+    query = f""" 
+        SELECT *
+        FROM concept_ancestor
+        WHERE ancestor_concept_id IN(:root_cids)
+          AND descendant_concept_id IN(:cids)
+          AND min_levels_of_separation > 0
+        ORDER BY ancestor_concept_id, min_levels_of_separation
+        """
+    query = query.replace(':root_cids', root_cids).replace(':cids', cids)
+    relationships: List[Dict] = [dict(x) for x in sql_query(con, query, return_with_keys=True)]
+    direct_relationships: List[Dict] = [
+        x for x in relationships if x['min_levels_of_separation'] == 1 and x['max_levels_of_separation'] == 1]
+    parent_children_map: Dict[int, List[int]] = {}
+    for x in direct_relationships:
+        parent_children_map.setdefault(x['ancestor_concept_id'], []).append(x['descendant_concept_id'])
+    return parent_children_map
+
+
+def hierarchy(root_cids: List[int], selected_concept_ids: List[int]) -> (Dict[int, Union[Dict, None]], List[int]):
```

Comment:
@Sigfried This function `hierarchy()` is almost the same. Most of the work to solve the issues we had was handled by your `new_hiearchy()` function (renamed to `hierarchy_helper()` earlier and now called `get_parent_children_map()`.

The most major change is that it returns the list of orphans.